nl_server/config_reader.py

```diff
@@ -0,0 +1,215 @@
+    if os.path.exists(full_user_path):
+      all_paths.append(full_user_path)
+
+  # Now load and merge all the catalog config files
+  catalog = Catalog(
+      version='',
+      indexes={},
+      models={},
+  )
+  for p in all_paths:
+    logging.info('Loading index and model catalog from: %s', p)
+    with open(p) as f:
+      partial_catalog = yaml.safe_load(f.read())
+      # read version
+      ver = catalog.version
+      this_ver = partial_catalog['version']
+      if (ver and this_ver and ver != this_ver):
+        raise ValueError(
+            f'Inconsistent version in config file {p}: {ver} and {this_ver}')
+
+      if this_ver not in _SUPPORTED_VERSIONS:
+        raise ValueError(f'Unsupported version: {this_ver} in file {p}')
+      catalog.version = this_ver
+
+      indexes = catalog.indexes
+      models = catalog.models
+
+      # TODO: when another version is added, extract separate version read as
+      # a function.
+      if this_ver == 1:
+        # read indexes
+        for index_name, index_config in partial_catalog['indexes'].items():
+          store_type = index_config['store_type']
+          match store_type:
+            case StoreType.MEMORY:
+              indexes[index_name] = MemoryIndexConfig(**index_config)
+            case StoreType.LANCEDB:
+              indexes[index_name] = LanceDBIndexConfig(**index_config)
+            case StoreType.VERTEXAI:
+              indexes[index_name] = VertexAIIndexConfig(**index_config)
+            case _:
+              raise ValueError(f'Unknown store type: {store_type}')
+
+          # read models
+          for model_name, model_config in partial_catalog['models'].items():
+            model_type = model_config['type']
+            match model_type:
+              case ModelType.LOCAL:
+                models[model_name] = LocalModelConfig(**model_config)
+              case ModelType.VERTEXAI:
+                models[model_name] = VertexAIModelConfig(**model_config)
+              case _:
+                raise ValueError(f'Unknown model type: {model_type}')
+  logging.debug('NL Catalog:\n%s', json.dumps(asdict(catalog), indent=2))
+  return catalog
+
+
+def read_env() -> Env:
+  """Reads the envs
+  """
+  # TODO: Make this generic by checking a user provided config file path
+  if custom_dc_util.is_custom_dc():
+    c = Env.from_dict(yaml.safe_load(open(_ENV_USER_PATH)))
+  elif os.path.exists(_ENV_MOUNT_PATH):
+    c = Env.from_dict(yaml.safe_load(open(_ENV_MOUNT_PATH)))
+  else:
+    with open(_ENV_CODE_PATH) as f:
+      full_nl_config = yaml.safe_load(f.read())
+      c = Env.from_dict(full_nl_config['nl']['env'])
+  logging.debug('NL Env:\n%s', json.dumps(asdict(c), indent=2))
+  return c
+
+
+def get_server_config(catalog: Catalog, env: Env) -> ServerConfig:
+  """
+  Merges the catalog and env into a server config.
+  """
+  default_indexes = [x for x in env.default_indexes if x in catalog.indexes]
+
+  # Add reranking models
+  models = {}
```

Comment:
Could you add type info?

Reply:
Or, start with empty ServerConfig, and fill it up, so types are clear

Reply:
aha, i moved that to this formation from last round of comment. I added typing for the struct now.